rust/src/io/exec/knn.rs

```diff
@@ -0,0 +1,186 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use arrow_array::cast::as_struct_array;
+use arrow_array::{ArrayRef, RecordBatch, StructArray};
+use arrow_ord::sort::sort_to_indices;
+use arrow_schema::{DataType, Field as ArrowField};
+use arrow_select::concat::concat_batches;
+use arrow_select::take::take;
+use futures::stream::{self, Stream, StreamExt, TryStreamExt};
+use tokio::sync::mpsc::Receiver;
+use tokio::task::JoinHandle;
+
+use super::{ExecNode, NodeType};
+use crate::arrow::*;
+use crate::index::vector::Query;
+use crate::utils::distance::l2_distance;
+use crate::{Error, Result};
+
+/// KNN node for post-filtering.
+pub struct KNNFlat {
+    rx: Receiver<Result<RecordBatch>>,
+
+    _bg_thread: JoinHandle<()>,
+}
+
+impl KNNFlat {
+    pub(crate) fn new(
+        child: impl Stream<Item = Result<RecordBatch>> + Unpin + Send + 'static,
+        query: &Query,
+    ) -> Self {
+        let (tx, rx) = tokio::sync::mpsc::channel(2);
+
+        let q = query.clone();
+        let bg_thread = tokio::spawn(async move {
+            let batches = child
+                .zip(stream::repeat_with(|| q.clone()))
+                .then(|(batch, q)| async move {
+                    let batch = batch?;
+                    let vectors = batch.column_with_name(&q.column).unwrap().clone();
```

Comment:
A lot of mixing of ? And unwraps. Feels like should be using ? As much as possible here